src/facade/cmd_arg_parser.h

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <optional>
+#include <string_view>
+
+#include "facade/facade_types.h"
+
+namespace facade {
+
+// Utility class for easily parsing command options from argument lists.
+struct CmdArgParser {
+  enum ErrorType {
+    OUT_OF_BOUNDS,
+    SHORT_OPT_TAIL,
+    INVALID_INT,
+    INVALID_CASES,
+  };
+
+  struct NextProxy;
+
+  template <typename T> struct CaseProxy {
+    operator T();
+
+    CaseProxy Case(std::string_view tag, T value);
+
+   private:
+    friend struct NextProxy;
+
+    CaseProxy(CmdArgParser* parser, size_t idx) : parser_{parser}, idx_{idx} {
+    }
+
+    CmdArgParser* parser_;
+    size_t idx_;
+    std::optional<T> value_;
+  };
+
+  struct NextProxy {
+    operator std::string_view() {
+      return parser_->SafeSV(idx_);
+    }
+
+    operator std::string() {
+      return std::string{operator std::string_view()};
+    }
+
+    template <typename T> T Int();
+
+    // Detect value based on cases.
+    // Returns default if the argument is not present among the cases list,
+    // and reports an error.
+    template <typename T> auto Case(std::string_view tag, T value) {
+      return CaseProxy<T>{parser_, idx_}.Case(tag, value);
+    }
+
+   private:
+    friend struct CmdArgParser;
+
+    NextProxy(CmdArgParser* parser, size_t idx) : parser_{parser}, idx_{idx} {
+    }
+
+    CmdArgParser* parser_;
+    size_t idx_;
+  };
+
+  struct CheckProxy {
+    explicit operator bool() const;
+
+    // Expect the tag to be followed by a number of arguments.
+    // Reports an error if the tag is matched but the condition is not met.
+    CheckProxy& ExpectTail(size_t tail) {
+      expect_tail_ = tail;
+      return *this;
+    }
+
+    // Call ToUpper on the next value after the flag and its expected tail.
+    CheckProxy& NextUpper() {
+      next_upper_ = true;
+      return *this;
+    }
+
+    CheckProxy& IgnoreCase() {
+      ignore_case_ = true;
+      return *this;
+    }
+
+   private:
+    friend struct CmdArgParser;
+
+    CheckProxy(CmdArgParser* parser, std::string_view tag, size_t idx)
+        : parser_{parser}, tag_{tag}, idx_{idx} {
+    }
+
+    CmdArgParser* parser_;
+    std::string_view tag_;
+    size_t idx_;
+    size_t expect_tail_ = 0;
+    bool next_upper_ = false;
+    bool ignore_case_ = false;
+  };
+
+  struct ErrorInfo {
+    ErrorType type;
+    size_t index;
+
+    ErrorReply MakeReply() const;
+  };
+
+ public:
+  CmdArgParser(CmdArgList args) : args_{args} {
+  }
+
+  // Debug asserts sure error was consumed
+  ~CmdArgParser();
+
+  // Get next value without consuming it
+  NextProxy Peek() {
+    return NextProxy(this, cur_i_);
+  }
+
+  // Consume next value
+  NextProxy Next() {
+    if (cur_i_ >= args_.size())
+      Report(OUT_OF_BOUNDS, cur_i_);
+    return NextProxy{this, cur_i_++};
+  }
```

Comment:
These two are the same with the only different that `Next` also reports an error if we go out of bounds. There is also a bug in `Next()`. 

If you chain `Next` from a single call you end up in the same id. `parser.Next().Next().Next()`

So: 
```
parser = make(); //cur_i == 0
parser.Next(). // Returns a NextProxy{this, 0} 
      .Next(). // Returns a NextProxy{this, 0}
      .Next()  //Returns a NextProxy{this, 0}
```
The `cur_i_++` is a bug.

I would consider decoupling the start of parsing, via a `Start` method and then start a chaining sequence via the `Next's` methods

Also add a test for this so we don't run into this problem again :)

Reply:
You can't call Next() from a NextProxy 😝 

Peek is though to be a "safe" lookup. For example `while (parser.Peek().Int<int64_t>() > 0) parser.Skip(1)` will skip all integers without accidentally running out of bounds.